### ISSUE TITLE
Pair type is not recognized for miniwdl executor

### ISSUE THREAD

**Issue by przemekiljan** [EliLillyCo/pytest-wdl#144]
While loading input files, during testing, pytest-wdl returned following message:
```
============================= test session starts ==============================
platform linux -- Python 3.8.3, pytest-5.3.5, py-1.9.0, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /builds/przemekk/16s-metagenomic-analysis
plugins: subtests-0.3.2, wdl-1.4.0
collecting ... 
collected 1 item

tests/test_16S_analysis.py::test_16S_analysis FAILED

=================================== FAILURES ===================================
______________________________ test_16S_analysis _______________________________

workflow_data = <pytest_wdl.core.DataManager object at 0x7fd0ecb15160>
workflow_runner = <pytest_wdl.fixtures.WorkflowRunner object at 0x7fd0ecb15100>

    def test_16S_analysis(workflow_data, workflow_runner):
        inputs = {
            "input_samples": workflow_data["samples"],
            "metadata": workflow_data["metadata"],
            "ref_base": workflow_data["ref_base"],
        }
        expected = workflow_data.get_dict("feature_table")
>       workflow_runner(
            "test_analysis.wdl",
            inputs,
            expected
        )

tests/test_16S_analysis.py:8: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.8/site-packages/pytest_wdl/fixtures.py:310: in __call__
    outputs[executors[0]] = self._run_test(executors[0], **call_args)
/usr/local/lib/python3.8/site-packages/pytest_wdl/fixtures.py:398: in _run_test
    outputs = executor.run_workflow(
/usr/local/lib/python3.8/site-packages/pytest_wdl/executors/miniwdl.py:78: in run_workflow
    target, input_env, input_json = CLI.runner_input(
/usr/local/lib/python3.8/site-packages/WDL/CLI.py:601: in runner_input
    input_env = runner_input_json_file(
/usr/local/lib/python3.8/site-packages/WDL/CLI.py:695: in runner_input_json_file
    ans = values_from_json(input_json, available_inputs, namespace=namespace)
/usr/local/lib/python3.8/site-packages/WDL/__init__.py:195: in values_from_json
    ans = ans.bind(key2, Value.from_json(ty, values_json[key]))
/usr/local/lib/python3.8/site-packages/WDL/Value.py:378: in from_json
    return Array(type, [from_json(type.item_type, item) for item in value])
/usr/local/lib/python3.8/site-packages/WDL/Value.py:378: in <listcomp>
    return Array(type, [from_json(type.item_type, item) for item in value])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

type = <WDL.Type.Pair object at 0x7fd0ec3fe040>
value = {'left': 'mock20', 'right': ['mock20-forward-read.fastq.gz', 'mock20-reverse-read.fastq.gz']}

    def from_json(type: Type.Base, value: Any) -> Base:
        """
        Instantiate a WDL value of the specified type from a parsed JSON value (str, int, float, list, dict, or null).
    
        :raise WDL.Error.InputError: if the given value isn't coercible to the specified type
        """
        if isinstance(type, (Type.Boolean, Type.Any)) and value in [True, False]:
            return Boolean(value)
        if isinstance(type, (Type.Int, Type.Any)) and isinstance(value, int):
            return Int(value)
        if isinstance(type, (Type.Float, Type.Any)) and isinstance(value, (float, int)):
            return Float(float(value))
        if isinstance(type, Type.File) and isinstance(value, str):
            return File(value)
        if isinstance(type, (Type.String, Type.Any)) and isinstance(value, str):
            return String(value)
        if isinstance(type, Type.Array) and isinstance(value, list):
            return Array(type, [from_json(type.item_type, item) for item in value])
        if (
            isinstance(type, Type.Map)
            and type.item_type[0] == Type.String()
            and isinstance(value, dict)
        ):
            items = []
            for k, v in value.items():
                assert isinstance(k, str)
                items.append((from_json(type.item_type[0], k), from_json(type.item_type[1], v)))
            return Map(type.item_type, items)
        if (
            isinstance(type, Type.StructInstance)
            and isinstance(value, dict)
            and type.members
            and set(type.members.keys()) == set(value.keys())
        ):
            items = {}
            for k, v in value.items():
                assert isinstance(k, str)
                items[k] = from_json(type.members[k], v)
            return Struct(Type.Object(type.members), items)
        if type.optional and value is None:
            return Null()
>       raise Error.InputError(
            "couldn't construct {} from input {}".format(str(type), json.dumps(value))
        )
E       WDL.Error.InputError: couldn't construct Pair[String,Array[File]] from input {"left": "mock20", "right": ["mock20-forward-read.fastq.gz", "mock20-reverse-read.fastq.gz"]}

/usr/local/lib/python3.8/site-packages/WDL/Value.py:402: InputError
============================== 1 failed in 0.80s ===============================
section_end:1597834063:build_script
[0Ksection_start:1597834063:after_script
[0K[0K[36;1mRunning after_script[0;m
[0;msection_end:1597834065:after_script
[0Ksection_start:1597834065:upload_artifacts_on_failure
[0K[0K[36;1mUploading artifacts for failed job[0;m
[0;msection_end:1597834067:upload_artifacts_on_failure
[0K[31;1mERROR: Job failed: exit code 1
[0;m
```  

As you can see `{"left": "mock20", "right": ["mock20-forward-read.fastq.gz", "mock20-reverse-read.fastq.gz"]}` is a valid WDL type of `Pair[String,Array[File]]`.  Error message just filled in gaps in: 
```
raise Error.InputError(
            "couldn't construct {} from input {}".format(str(type), json.dumps(value))
        )
```  
I recognized it as a bug when I couldn't find Pair type in `isinstance` declarations.  
I understand if fault is on my side, and would accept any help or provide further information about my setup. I have relatively complicated format for input samples: `Array[Pair[String,Array[File]]]`, and simplifying it is not possible.  

Just to be clear I will paste here snippets from .json and python files that might be useful:  

JSON file that was loaded as a file with parameters to WDL to test it, before using pytest-wdl:
```
{
  "test_analysis.input_samples": [{"left":"mock.20",
                                    "right": ["mock20-forward-read.fastq.gz",
                                              "mock20-reverse-read.fastq.gz"]},
                                {"left":"mock.21",
                                    "right": ["mock21-forward-read.fastq.gz",
                                              "mock21-reverse-read.fastq.gz"]}]
}
```
Those samples are interpreted in workflow, inside .wdl as `Array[Pair[String,Array[File]]] input_samples`  

Here is test_data.json:  
```
{
  "samples": [{"left":"mock20",
              "right": ["mock20-forward-read.fastq.gz",
                        "mock20-reverse-read.fastq.gz"]},
              {"left":"mock21",
              "right": ["mock21-forward-read.fastq.gz",
                        "mock21-reverse-read.fastq.gz"]}],
  "feature_table": {
    "name" : "feature_table_filtered.tsv"
  }
}
```
and it is interpreted in test_analysis.py (same as test_variant_caller.py in quick start):
```
def test_analysis(workflow_data, workflow_runner):
    inputs = {
        "input_samples": workflow_data["samples"]
    }
    expected = workflow_data.get_dict("feature_table")
    workflow_runner(
        "test_analysis.wdl",
        inputs,
        expected
    )
```

I also thought about downloading samples by using `"url":<url to sample>` that was an example in your quick start guide, but I wouldn't even know where to put it in such complicated structure. 

Also, I have a question - is there a way to globally specify docker image that tests should use? I've done it through  `"default_runtime_attributes"` option, before using pytest-wdl.

**Comment by michaeltneylon** [comment]
I can reproduce this, but only when using the default executor `miniwdl`. This is either a limitation of `miniwdl` itself, the implementation here of how that type is being converted to `miniwdl`, or perhaps only specific to the older version of `miniwdl` used in this version `pytest-wdl`.  I am going to update the issue title to reflect this is specific to miniwdl executor.

In the meantime, I recommend using the `cromwell` executor as it does support `Pair` type.

**Comment by przemekiljan** [comment]
Indeed, changing executor fixed my problem. But I've encountered another problem. I want to avoid hardcoding so I can run same tests for both my local environment as well as gitlab CI/CD platform. While using "path" key for other arguments worked well, that so I can insert ${PWD} in there, I cannot think of a way to specify path to files mentioned in test_data.json, in key "samples" in the same way. I've tried:
```
{
  "samples": [{"left":"mock20",
              "right": [{"path": "mock20-forward-read.fastq.gz"},
                        {"path": "mock20-reverse-read.fastq.gz"}]},
              {"left":"mock21",
              "right": [{"path": "mock21-forward-read.fastq.gz"},
                        {"path": "mock21-reverse-read.fastq.gz"}]}],
  "metadata": {"path": "${PWD}/tests/data/metadata.tsv"}
}
```
but I does not seem to recognize that those are indeed paths to my samples. What should I do to properly specify paths to my samples while retaining my `Array[Pair[String,Array[File]]]` scheme?
